### ISSUE TITLE
Reprise: Error after deployment Only "Cannot read properties of undefined (reading 'id')"

### ISSUE THREAD

**Issue by tandrasi** [SharePoint/sp-dev-docs#9548]
### Target SharePoint environment

SharePoint Online

### What SharePoint development model, framework, SDK or API is this about?

💥 SharePoint Framework

### Developer environment

Windows

### What browser(s) / client(s) have you tested

- [ ] 💥 Internet Explorer
- [X] 💥 Microsoft Edge
- [ ] 💥 Google Chrome
- [ ] 💥 FireFox
- [ ] 💥 Safari
- [ ] mobile (iOS/iPadOS)
- [ ] mobile (Android)
- [ ] not applicable
- [ ] other (enter in the "Additional environment details" area below)

### Additional environment details

- browser version latest
- SPFx version 1.18.2
- Node.js version 18.18.0

### Describe the bug / error

**TL;DR**
I'm sorry for the verbosity below, but our situation is quite complex and I'm trying my best to ensure everything is relayed as best as I can, but here's the crux of the issue: **using specific dev dependencies in our package.json causes debug, build, and package to work with our SPFx app, but when deployed, app fails to load from app catalog with an _TypeError: Cannot read properties of undefined (reading 'id')_ error.**

**Preface**
We have internal frameworks which are packaged up as NPM packages on our internal feed. We use these as baseline dependencies for all our SPFx projects as they contain a plethora of code just ready to go and use. For simplicity's sake, let's call our frameworks "shared", "forms-shared", and "reports-shared". 

This is the hierarchy - basically, "shared" is the foundation that everything else is built on - even our other NPM packages:

- "shared"
---- "forms-shared"
---- "reports-shared"

All our SPFx projects use "shared" as a base dependency, but have the option of using "forms-shared" or "reports-shared" depending on what we're building the SPFx App for.

**Journey thus far**
1. I was upgrading our "forms-shared" project with new functionality and bug fixes. Also, some dependencies were upgraded like going from SPFx 1.18.0 to 1.18.2. 
2. After this was built and published to our Azure Artifacts feed, I pull in the new NPM package to my actual SPFx web part. 
3. I ran debug and it all worked fine. I could see all new changes and was happy to package and deploy.
4. I packaged the SPFx project for deployment and pushed it to our Site Collection level App Catalog.
5. When trying to load the app, I was greeted with the following error message:

```
CALL STACK:
TypeError: Cannot read properties of undefined (reading 'id')
    at https://res-1.cdn.office.net/files/sp-client/sp-pages-assembly_en-us_3d3eb7503bf5891494467738f00140f9.js:70:747315
    at new Promise (<anonymous>)
    at e.loadComponent (https://res-1.cdn.office.net/files/sp-client/sp-pages-assembly_en-us_3d3eb7503bf5891494467738f00140f9.js:70:747283)
    at https://res-1.cdn.office.net/files/sp-client/sp-pages-assembly_en-us_3d3eb7503bf5891494467738f00140f9.js:70:749044
    at async Promise.all (index 0)
```

I saw that something similar was posted here #8934 and was convinced it was something to do with dependencies.

**Note:** I tried this on multiple tenancies and between Tenancy and Site Collection level app catalogs, but still got the same issue.

### Steps to reproduce

Ok, now that the description is out of the way and after two full days trying to debug this issue, I found what it was and have successfully got it working again, **but I don't quite agree that it's working as intended**. It was indeed a dependency issue but the strange thing is that it worked like this before I did some upgrades.

The SPFx app in question was using our "shared" and "forms-shared" frameworks. In order of dependency hierarchy, this is what it looks like:

-> App
--> "shared"
**----> "forms-shared"**

**_I singled out that it was "forms-shared" using "shared" as a dev dependency that was causing the issue._**

### **What Doesn't Work with "forms-shared"**
"forms-shared", as stated before, is dependent on "shared" being installed. **However, since all our SPFx projects use "shared" as a base dependency, this is what the erroneous "forms-shared" package.json looked like when I was experiencing the issues on deploy:**

 ```
forms-shared package.json

"dependencies": {
    "@fluentui/react": "^8.115.6",
    ...
  },
  "devDependencies": {
    "@jch/spfx-shared": "^2.0.24",
    "@microsoft/eslint-config-spfx": "1.18.2",
    ...
  }
```

As you can see, **@jch/spfx-shared**, is a dev dependency as I needed on the typings and I knew that internally, all our SPFx apps have "shared" as a base dependency, so there is no need to double down on compile times (it does make a difference in our scenario). **This has worked for us for the longest time, but I think due to some updates/upgrades in dependencies between projects, the App Catalog just simply refused to process and cache the app in the CDN.**

### **What Works with "forms-shared"**
Now, to **fix the issue and make it work when I deploy**, all that needed to be done is to move **@jch/spfx-shared** as a dependency, rather than a dev dependency.

 ```
forms-shared package.json

"dependencies": {
    "@fluentui/react": "^8.115.6",
    "@jch/spfx-shared": "^2.0.24",
    ...
  },
  "devDependencies": {
    "@microsoft/eslint-config-spfx": "1.18.2",
    ...
  }
```

### **FYI - App package.json**
My main App's package.json looks like this:
```
app package.json

"dependencies": {
    "@fluentui/react": "^8.115.6",
    "@jch/spfx-shared": "^2.0.24",
    "@jch/spfx-forms-shared": "^1.0.15",
    ...
  },
  "devDependencies": {
    "@microsoft/eslint-config-spfx": "1.18.2",
    ...
```

### Expected behavior

Semantically and syntactically, having our "shared" framework as a dev dependency on our "forms-shared" framework should work on deployment as it works in debug. All the correct packages exist during build and package.

**Comment by iseliweb** [comment]
Got the same error message when integrating a custom developed property pane dropdown for my SPFX SharePoint webpart.
After downgrading to react@17.0.1 (from 17.0.2) it started working again.

` npm i react@17.0.1 react-dom@17.0.1 --save`

As there are also a lot of other projects affected with same error message but different code, this should be addressed soon in the react package.

**Comment by AJIXuMuK** [comment]
@tandrasi - sorry for the long delay and thank you for the detailed explanation.
Dealing with dependencies could be a quite huge problem, especially with shared ones.
I would suggest having either a strict version used or have some tools (like dedupe mentioned above) to keep dependencies in sync.
Also, with SPFx 1.19 we transitioned to Webpack 5 that should be a bit better with bundling.

Is there anything else you feel like should be done from our side? 
Thanks!

**Comment by tandrasi** [comment]
@AJIXuMuK, thanks for getting back to me. What we've done is consolidate all our frameworks into one, so we no longer face this issue.

As for using SPFx 1.19 - this would be great but I get other issues such as [Object object] errors which I've already reported here: https://github.com/SharePoint/sp-dev-docs/issues/9675 (which was closed) with another one opened by another user here: https://github.com/SharePoint/sp-dev-docs/issues/9834 (which I've seen you've already labelled). This would be awesome to get fixed so I can try some other library I have in the works!

Thanks again

**Comment by AJIXuMuK** [comment]
@tandrasi - the closed issue has a workaround on how to get the actual error message. And we will fix that - to show the error message correctly. 
But the actual issue is with some imports in your code.

**Comment by tandrasi** [comment]
@AJIXuMuK yes, from my code, it seems to be @pnp/spfx-property-controls but am reluctant to try the v4 beta on production, so I'll wait for a fix :). Cheers!

**Comment by AJIXuMuK** [comment]
There is a v3 beta for controls. And we will make a release this week

**Comment by Amey-MSFT** [comment]
Hello @tandrasi,
Thank you for bringing this issue to our attention. Could you please confirm if the issue still persists for you?
We will look into it and get back to you shortly.

**Comment by Amey-MSFT** [comment]
We are closing this issue for now. If the problem persists, feel free to reopen it or open a new one.
